### PR TITLE
Refactor out states into separate classes

### PR DESCRIPTION
**Pull request by planetguy32**
Also moves code generally into classes per movement state. (There are still scattered places outside State* that behave differently per state, but much of the state-specific code is collected. Client-side code is generally ignored in this refactor.)

Having code for each movement state placed based on the movement state makes more sense to me - if you disagree, feel free to close the PR.

### CODE REVIEW

**Comment by elveskevtar** [comment]
I really like this organization actually, makes a lot more sense. I'll do some testing tonight and tomorrow and would like to move to this. Great work!

File: src/main/java/net/smart/moving/player/state/StateHandlerCrawl.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+import net.smart.moving.player.SmartBase;
+
+public class StateHandlerCrawl extends StateHandler {
+
+    private void applyHorizontalDamping(EntityPlayer player, float horizontalDamping) {
+        player.motionX *= horizontalDamping;
+        player.motionZ *= horizontalDamping;
+    }
+
+    public void afterOnUpdate(EntityPlayer player) {
```
Comment (elveskevtar) [inline src/main/java/net/smart/moving/player/state/StateHandlerCrawl.java:16]: Nitpick: add override annotation

File: src/main/java/net/smart/moving/player/state/StateHandlerCrawl.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        applyHorizontalDamping(player, PlayerConstants.CRAWL_DAMPING_FACTOR);
+        player.height = PlayerConstants.CRAWL_HEIGHT;
+        player.eyeHeight = PlayerConstants.CRAWL_EYE_HEIGHT;
+    }
+
+    public void updatePlayerRotations(EntityPlayer player, SmartBase smartBase) {
```
Comment (elveskevtar) [inline src/main/java/net/smart/moving/player/state/StateHandlerCrawl.java:23]: Nitpick: add override annotation

File: src/main/java/net/smart/moving/player/state/StateHandlerIdle.java [hunk 3]
```diff
@@ -0,0 +1,13 @@
+package net.smart.moving.player.state;
+
+import net.minecraft.entity.player.EntityPlayer;
+import net.smart.moving.player.PlayerConstants;
+
+public class StateHandlerIdle extends StateHandler {
+
+    public void afterOnUpdate(EntityPlayer player) {
```
Comment (elveskevtar) [inline src/main/java/net/smart/moving/player/state/StateHandlerIdle.java:9]: Nitpick: add override annotation

**Review by elveskevtar — CHANGES_REQUESTED** [review]
Tested out and it works great. This is a much better design pattern. I requested a few nitpicks but once those are fixed, it's good to merge!!

If you could rebase your commits into 1 commit when doing the revision that would be amazing. Still need to make that contributing page.

**Review by elveskevtar — APPROVED** [review]
Looks good to squash and merge!

**Comment by planetguy32** [comment]
I think I've set it up with the annotations you requested. I'm still not really Git literate, sorry about the bizarre state of things.

**Comment by elveskevtar** [comment]
That's okay! It's a little wonky but almost everything in git is fixable. Rebasing and squashing can be a little dangerous bc you're actually changing the repositories commit history so I would rather lean on the safer side than anything else.

**Pull request merged**